### ISSUE TITLE
Android tests are failing on Windows 10

### ISSUE THREAD

**Issue by hiyelbaz** [mozilla/geckodriver#1902]
I'm not sure but I think there is a problem with path generation on Windows.
In the logs, you can see the following path `/sdcardAndroid/dataorg.mozilla.firefox` which should be `/sdcard/Android/data/org.mozilla.firefox`.

## System
* Version: 0.29.1
* Driver Platform: Windows 10
* Firefox Platform: Android 10.0
* Firefox: (Android) 89.1.1
* Webdriverio: 7.7.3

## Testcase
```
let options = {
    hostname: 'localhost',
    port: 4444,
    path: '/',
    capabilities: {
        browserName: "firefox",
        acceptInsecureCerts: true,
        "moz:firefoxOptions": {
            androidPackage: "org.mozilla.firefox",
            //androidActivity: "org.mozilla.firefox.App",
            //androidDeviceSerial: "emulator-5554",
            log: { level: "TRACE" }
        },
    },
}
```

## Trace-level log
```
1624473749740   geckodriver     INFO    Listening on 127.0.0.1:4444
1624473799556   mozdevice       WARN    adb server response contained hexstring length 128 and message length was 128 and message was "emulator-5554          device product:sdk_gphone_x86_64 model:Android_SDK_built_for_x86_64 device:generic_x86_64 transport_id:3\n"
1624473801586   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473801586   mozdevice       DEBUG   execute_host_command: << []
1624473801587   mozdevice       DEBUG   execute_host_command: >> "shell:id"
1624473801600   mozdevice       DEBUG   execute_host_command: << "uid=2000(shell) gid=2000(shell) groups=2000(shell),1004(input),1007(log),1011(adb),1015(sdcard_rw),1028(sdcard_r),3001(net_bt_admin),3002(net_bt),3003(inet),3006(net_bw_stats),3009(readproc),3011(uhid) context=u:r:shell:s0\n"
1624473803632   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473803632   mozdevice       DEBUG   execute_host_command: << []
1624473803632   mozdevice       DEBUG   execute_host_command: >> "shell:su 0 id"
1624473803644   mozdevice       DEBUG   execute_host_command: << "/system/bin/sh: su: inaccessible or not found\n"
1624473805655   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473805656   mozdevice       DEBUG   execute_host_command: << []
1624473805656   mozdevice       DEBUG   execute_host_command: >> "shell:su -c id"
1624473805670   mozdevice       DEBUG   execute_host_command: << "/system/bin/sh: su: inaccessible or not found\n"
1624473805670   mozdevice       DEBUG   Device is unrooted
1624473807695   geckodriver::android    DEBUG   Android port forward (50431 -> 2829) started
1624473809752   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473809753   mozdevice       DEBUG   execute_host_command: << []
1624473809753   mozdevice       DEBUG   execute_host_command: >> "shell:echo $EXTERNAL_STORAGE"
1624473809763   mozdevice       DEBUG   execute_host_command: << "/sdcard\n"
1624473809763   geckodriver::android    DEBUG   Connecting: options=AndroidOptions { activity: Some("org.mozilla.fenix.IntentReceiverActivity"), device_serial: None, intent_arguments: Some(["-a", "android.intent.action.VIEW", "-d", "about:blank"]), package: "org.mozilla.firefox", storage: Auto }, storage=Sdcard) test_root=/sdcard\Android/data\org.mozilla.firefox\files/test_root, run_as_package=None
1624473811787   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473811788   mozdevice       DEBUG   execute_host_command: << []
1624473811788   mozdevice       DEBUG   execute_host_command: >> "shell:pm list packages org.mozilla.firefox"
1624473811817   mozdevice       DEBUG   execute_host_command: << "package:org.mozilla.firefox\n"
1624473813836   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473813836   mozdevice       DEBUG   execute_host_command: << []
1624473813837   mozdevice       DEBUG   execute_host_command: >> "shell:pm clear org.mozilla.firefox"
1624473813881   mozdevice       DEBUG   execute_host_command: << "Success\n"
1624473815915   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473815915   mozdevice       DEBUG   execute_host_command: << []
1624473815937   mozdevice       DEBUG   execute_host_command: << ""hell:pm grant org.mozilla.firefox android.permission.READ_EXTERNAL_STORAGE"
1624473817951   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473817952   mozdevice       DEBUG   execute_host_command: << []
1624473817952   mozdevice       DEBUG   execute_host_command: >> "shell:pm grant org.mozilla.firefox android.permission.WRITE_EXTERNAL_STORAGE"
1624473817972   mozdevice       DEBUG   execute_host_command: << ""
1624473817972   mozdevice       DEBUG   Creating /sdcard\Android/data\org.mozilla.firefox\files/test_root
1624473820000   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473820001   mozdevice       DEBUG   execute_host_command: << []
1624473820001   mozdevice       DEBUG   execute_host_command: >> "shell:mkdir -p /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root"
1624473820016   mozdevice       DEBUG   execute_host_command: << "mkdir: \'/sdcardAndroid\': Read-only file system\n"
1624473822063   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473822064   mozdevice       DEBUG   execute_host_command: << []
1624473822064   mozdevice       DEBUG   execute_host_command: >> "shell:chmod  -R 777 /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root"
1624473822076   mozdevice       DEBUG   execute_host_command: << "chmod: /sdcardAndroid/dataorg.mozilla.firefoxfiles/test_root: No such file or directory\n"
1624473822077   mozdevice       DEBUG   Deleting /sdcard\Android/data\org.mozilla.firefox\files/test_root\org.mozilla.firefox-geckodriver-profile
1624473824123   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473824124   mozdevice       DEBUG   execute_host_command: << []
1624473824124   mozdevice       DEBUG   execute_host_command: >> "shell:rm -rf /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root\\org.mozilla.firefox-geckodriver-profile"
1624473824137   mozdevice       DEBUG   execute_host_command: << ""
1624473824137   mozdevice       DEBUG   Pushing C:\Users\Burak\AppData\Local\Temp\rust_mozprofilehpMPdL to /sdcard\Android/data\org.mozilla.firefox\files/test_root\org.mozilla.firefox-geckodriver-profile
1624473826194   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473826195   mozdevice       DEBUG   execute_host_command: << []
1624473826195   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root\\org.mozilla.firefox-geckodriver-profile"
1624473826213   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid/dataorg.mozilla.firefoxfiles/test_rootorg.mozilla.firefox-geckodriver-profile: No such file or directory\n"
1624473828253   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473828254   mozdevice       DEBUG   execute_host_command: << []
1624473828254   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root"
1624473828270   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid/dataorg.mozilla.firefoxfiles/test_root: No such file or directory\n"
1624473830318   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473830319   mozdevice       DEBUG   execute_host_command: << []
1624473830319   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android/data\\org.mozilla.firefox\\files"
1624473830340   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid/dataorg.mozilla.firefoxfiles: No such file or directory\n"
1624473832380   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473832381   mozdevice       DEBUG   execute_host_command: << []
1624473832381   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android/data\\org.mozilla.firefox"
1624473832396   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid/dataorg.mozilla.firefox: No such file or directory\n"
1624473834439   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473834440   mozdevice       DEBUG   execute_host_command: << []
1624473834440   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android/data"
1624473834454   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid/data: No such file or directory\n"
1624473836498   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473836498   mozdevice       DEBUG   execute_host_command: << []
1624473836498   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard\\Android"
1624473836515   mozdevice       DEBUG   execute_host_command: << "ls: /sdcardAndroid: No such file or directory\n"
1624473838534   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473838535   mozdevice       DEBUG   execute_host_command: << []
1624473838535   mozdevice       DEBUG   execute_host_command: >> "shell:ls /sdcard"
1624473838554   mozdevice       DEBUG   execute_host_command: << "Alarms  DCIM     Movies Notifications Podcasts  \nAndroid Download Music  Pictures      Ringtones \n"
1624473838554   mozdevice       DEBUG   Creating /sdcard\Android/data\org.mozilla.firefox\files/test_root\org.mozilla.firefox-geckodriver-profile
1624473840587   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473840587   mozdevice       DEBUG   execute_host_command: << []
1624473840587   mozdevice       DEBUG   execute_host_command: >> "shell:mkdir -p /sdcard\\Android/data\\org.mozilla.firefox\\files/test_root\\org.mozilla.firefox-geckodriver-profile"
1624473840608   mozdevice       DEBUG   execute_host_command: << "mkdir: \'/sdcardAndroid\': Read-only file system\n"
1624473842643   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473842644   mozdevice       DEBUG   execute_host_command: << []
1624473842644   mozdevice       DEBUG   execute_host_command: >> "shell:chmod  -R 777 /sdcard\\Android"
1624473842660   mozdevice       DEBUG   execute_host_command: << "chmod: /sdcardAndroid: No such file or directory\n"
1624473846746   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473846747   mozdevice       DEBUG   execute_host_command: << []
1624473846747   mozdevice       DEBUG   execute_host_command: >> "shell:am clear-debug-app org.mozilla.firefox"
1624473846781   mozdevice       DEBUG   execute_host_command: << ""
1624473846781   geckodriver::android    DEBUG   Disabled reading from configuration file
1624473846781   mozdevice       DEBUG   Deleting /data/local/tmp/org.mozilla.firefox-geckoview-config.yaml
1624473848811   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473848812   mozdevice       DEBUG   execute_host_command: << []
1624473848812   mozdevice       DEBUG   execute_host_command: >> "shell:rm -rf /data/local/tmp/org.mozilla.firefox-geckoview-config.yaml"
1624473848821   mozdevice       DEBUG   execute_host_command: << ""
1624473848821   geckodriver::android    DEBUG   Deleted GeckoView configuration file
1624473850837   mozdevice       DEBUG   execute_host_command: >> "host:transport:emulator-5554"
1624473850838   mozdevice       DEBUG   execute_host_command: << []
1624473850838   mozdevice       DEBUG   execute_host_command: >> "killforward:tcp:50431"
1624473850839   geckodriver::android    ERROR   Android port forward (50431 -> 2829) failed to stop: adb error: closed
1624473850840   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"unknown error","message":"adb error: secure_mkdirs failed: Read-only file system","stacktrace":""}}
1624473850857   webdriver::server       DEBUG   -> POST /session {"capabilities":{"alwaysMatch":{"browserName":"firefox","acceptInsecureCerts":true,"moz:firefoxOptions":{"androidPackage":"org.mozilla.firefox","log":{"level":"TRACE"},"args":["-remote-debugging-port","50429"]}},"firstMatch":[{}]},"desiredCapabilities":{"browserName":"firefox","acceptInsecureCerts":true,"moz:firefoxOptions":{"androidPackage":"org.mozilla.firefox","log":{"level":"TRACE"},"args":["-remote-debugging-port","50429"]}}}
```

**Comment by whimboo** [comment]
Thanks for reporting. This issue is already tracked as https://bugzilla.mozilla.org/show_bug.cgi?id=1610629.

**Comment by whimboo** [comment]
Actually this is a newer regression as shipped with 0.29.1. As such I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1721374 to get it fixed for the next geckodriver release.

**Comment by whimboo** [comment]
@hiyelbaz would you mind trying out a nightly build of geckodriver and check if that makes it work for you? It can be found here:

https://treeherder.mozilla.org/jobs?repo=mozilla-central&searchStr=gd%2Cwindows&selectedTaskRun=aTPiD2qzQpa99-CbSqWy0w.0

In the lower pane you can see the `geckodriver.zip` file. Thanks a lot!

**Comment by whimboo** [comment]
Closing as this has been fixed and will be shipped with the upcoming 0.30.0 release.
